File: bindings/typescript/tests/server.test.ts
/**
 * Unit tests for Node broadcast deployment lifecycle (SCP-296).
 *
 * Tests mock the native addon; no Rust extension required.
 *
 * See spec section 18.11.8 and `.docs/prds/http-features.json` SCP-296.
 */

import { describe, expect, it, mock } from "bun:test";
import type { SiteConfig } from "../src/types";

// ---------------------------------------------------------------------------
// Minimal mock for NativeNodeHandle (bypasses native addon loading)
// ---------------------------------------------------------------------------

interface MockNodeHandle {
  relayUrl: string;
  relayPort: number;
  did: string;
  isShutdown: boolean;
  shutdown: () => void;
  enableSiteProjection: (
    contextId: string,
    admission: string,
    hostname: string,
    broadcastKeyHex: string | null,
    authorDid: string | null,
    indexPath: string | null,
    maxAssetsPerDeploy: number | null,
    maxDeploySizeBytes: number | null,
    deployRetentionCount: number | null,
    cspOverride: string | null,
  ) => Promise<void>;
  commitDeploy: (contextId: string, deployId: string) => Promise<number>;
  rollbackDeploy: (contextId: string, deployId: string) => Promise<void>;
  disableSiteProjection: (contextId: string) => Promise<void>;
}

function createMockHandle(): MockNodeHandle {
  return {
    relayUrl: "ws://127.0.0.1:9876/scp/v1",
    relayPort: 9876,
    did: "did:dht:z6MkTestNode",
    isShutdown: false,
    shutdown: mock(() => {}),
    enableSiteProjection: mock(async () => {}),
    commitDeploy: mock(async () => 42),
    rollbackDeploy: mock(async () => {}),
    disableSiteProjection: mock(async () => {}),
  };
}

// The Node class uses private #handle, so we test through the public API
// by constructing Node with a mock handle via the private constructor.
// Since the constructor is private, we use a workaround: create a test
// subclass or access the class through the module's internal mechanism.
//
// Actually, the Node constructor IS private. We need to test through the
// server module. Let's import and test the Node class by constructing it
// through a mock pattern that accesses the private handle.

// Since the Node class in server.ts has a private constructor and relies on
// native addon loading, we test the lifecycle method signatures and validation
// at the type level and through integration patterns.

describe("Node lifecycle methods (SCP-296)", () => {
  // We test the method signatures exist and the type constraints are correct.
  // Full integration tests require a running native addon (see real-napi.test.ts).

  it("SiteConfig validation rejects empty hostname", () => {
    const { validateSiteConfig } = require("../src/types");
    expect(() => validateSiteConfig({ hostname: "" })).toThrow("hostname must not be empty");
  });

  it("SiteConfig validation rejects invalid retention count", () => {
    const { validateSiteConfig } = require("../src/types");
    expect(() => validateSiteConfig({ hostname: "example.com", deployRetentionCount: 0 })).toThrow(
      "deployRetentionCount must be an integer between 1 and 8",
    );
    expect(() => validateSiteConfig({ hostname: "example.com", deployRetentionCount: 9 })).toThrow(
      "deployRetentionCount must be an integer between 1 and 8",
    );
  });

  it("SiteConfig validation rejects unsafe CSP", () => {
    const { validateSiteConfig } = require("../src/types");
    expect(() =>
      validateSiteConfig({
        hostname: "example.com",
        cspOverride: "script-src 'unsafe-eval'",
      }),
    ).toThrow("CSP must not contain 'unsafe-eval'");
  });

  it("SiteConfig validation accepts valid config", () => {
    const { validateSiteConfig } = require("../src/types");
    expect(() =>
      validateSiteConfig({
        hostname: "mysite.example.com",
        indexPath: "/app.html",
        maxAssetsPerDeploy: 5000,
        maxDeploySizeBytes: 100_000_000,
        deployRetentionCount: 4,
        cspOverride: "default-src 'self'",
      }),
    ).not.toThrow();
  });

  it("SiteConfig validation accepts minimal config", () => {
    const { validateSiteConfig } = require("../src/types");
    expect(() => validateSiteConfig({ hostname: "example.com" })).not.toThrow();
  });

  it("Node class exports from server module", async () => {
    // Verify the Node class is exported. Phase 4 PR 4 (#1549, ADR-048)
    // deleted the static `Node.startInMemory` / `Node.startLocal`
    // factories — callers construct nodes via
    // `scp.nodeStartInMemory(...)` / `scp.nodeStartLocal(...)` which
    // hydrate via `Node._fromHandle`. The class itself is still a
    // real export used as a type and for the instance method surface.
    const { Node } = await import("../src/server");
    expect(Node).toBeDefined();
    expect(typeof Node._fromHandle).toBe("function");
  });

  it("Node class exports from index", async () => {
    // Verify Node is re-exported from the package root
    const { Node } = await import("../src/index");
    expect(Node).toBeDefined();
  });

  // Direct mock-based tests using the internal handle pattern
  it("enableSiteProjection delegates to native handle", async () => {
    const handle = createMockHandle();
    // We cannot construct Node directly (private constructor), but we
    // verify the mock handle interface matches the expected contract.
    const config: SiteConfig = {
      hostname: "mysite.example.com",
      indexPath: "/app.html",
      maxAssetsPerDeploy: 5000,
      maxDeploySizeBytes: 100_000_000,
      deployRetentionCount: 4,
      cspOverride: "default-src 'self'",
    };

    await handle.enableSiteProjection(
      "ctx-123",
      "open",
      config.hostname,
      "ab".repeat(32),
      "did:dht:z6MkAuthor",
      config.indexPath ?? null,
      config.maxAssetsPerDeploy ?? null,
      config.maxDeploySizeBytes ?? null,
      config.deployRetentionCount ?? null,
      config.cspOverride ?? null,
    );

    expect(handle.enableSiteProjection).toHaveBeenCalledTimes(1);
  });

  it("commitDeploy returns asset count", async () => {
    const handle = createMockHandle();
    const count = await handle.commitDeploy("ctx-123", "deploy-abc");
    expect(count).toBe(42);
  });

  it("commitDeploy propagates errors", async () => {
    const handle = createMockHandle();
    handle.commitDeploy = mock(async () => {
      throw new Error("not projected");
    });

    await expect(handle.commitDeploy("ctx-bad", "deploy-xyz")).rejects.toThrow("not projected");
  });

  it("rollbackDeploy delegates to native handle", async () => {
    const handle = createMockHandle();
    await handle.rollbackDeploy("ctx-123", "deploy-old");
    expect(handle.rollbackDeploy).toHaveBeenCalledTimes(1);
  });

  it("rollbackDeploy propagates errors", async () => {
    const handle = createMockHandle();
    handle.rollbackDeploy = mock(async () => {
      throw new Error("deploy not found");
    });

    await expect(handle.rollbackDeploy("ctx-bad", "deploy-nope")).rejects.toThrow(
      "deploy not found",
    );
  });

  it("disableSiteProjection delegates to native handle", async () => {
    const handle = createMockHandle();
    await handle.disableSiteProjection("ctx-123");
    expect(handle.disableSiteProjection).toHaveBeenCalledTimes(1);
  });

  it("disableSiteProjection is idempotent (no error on unprojected)", async () => {
    const handle = createMockHandle();
    // Should not throw — mock returns void.
    await handle.disableSiteProjection("ctx-nonexistent");
  });
});

File: bindings/kotlin/scp-kt/src/test/kotlin/works/limn/scp/ServerTest.kt
// ServerTest.kt -- Unit tests for Node broadcast deployment lifecycle (SCP-296)
//
// Tests use stub ServerBindings; no Rust binary required.
//
// Provenance: spec section 18.11.8, SCP-296

package works.limn.scp

import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestDispatcher
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import works.limn.scp.bridge.BridgeException
import works.limn.scp.bridge.CoroutineBridge
import works.limn.scp.conformance.ConformanceStubBindings
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ServerTest {

    private lateinit var testDispatcher: TestDispatcher
    private lateinit var stubBindings: StubServerBindings
    private lateinit var bridge: CoroutineBridge
    private lateinit var serverBridge: ServerBridge

    @BeforeEach
    fun setUp() {
        testDispatcher = StandardTestDispatcher()
        stubBindings = StubServerBindings()
        // CoroutineBridge requires NativeBindings — use ConformanceStubBindings.
        bridge = CoroutineBridge(
            nativeBindings = ConformanceStubBindings(),
            ioDispatcher = testDispatcher,
            cpuDispatcher = testDispatcher,
        )
        serverBridge = ServerBridge(bindings = stubBindings, bridge = bridge)
    }

    private suspend fun createNode(): Node {
        stubBindings.nodeStartInMemoryResult =
            """{"relayUrl":"ws://127.0.0.1:9876/scp/v1","relayPort":9876,"did":"did:dht:z6MkTestNode"}"""
        return Node.startInMemory(serverBridge)
    }

    // MARK: - enableSiteProjection

    @Test
    fun `enableSiteProjection delegates to bindings with correct arguments`() = runTest(testDispatcher) {
        val node = createNode()
        val config = SiteConfig(
            hostname = "mysite.example.com",
            indexPath = "/app.html",
            maxAssetsPerDeploy = 5000,
            maxDeploySizeBytes = 100_000_000L,
            deployRetentionCount = 4,
            cspOverride = "default-src 'self'",
        )

        node.enableSiteProjection(
            contextId = "ctx-123",
            admission = "open",
            config = config,
            broadcastKeyHex = "ab".repeat(32),
            authorDid = "did:dht:z6MkAuthor",
        )

        val args = stubBindings.lastEnableSiteProjectionArgs
        assertEquals("ctx-123", args?.contextId)
        assertEquals("ab".repeat(32), args?.broadcastKeyHex)
        assertEquals("did:dht:z6MkAuthor", args?.authorDid)
        assertEquals("open", args?.admission)
        assertEquals("mysite.example.com", args?.hostname)
        assertEquals("/app.html", args?.indexPath)
        assertEquals(5000, args?.maxAssetsPerDeploy)
        assertEquals(100_000_000L, args?.maxDeploySizeBytes)
        assertEquals(4, args?.deployRetentionCount)
        assertEquals("default-src 'self'", args?.cspOverride)
    }

    @Test
    fun `enableSiteProjection passes null for default SiteConfig values`() = runTest(testDispatcher) {
        val node = createNode()
        val config = SiteConfig(hostname = "example.com")

        node.enableSiteProjection(
            contextId = "ctx-456",
            admission = "gated",
            config = config,
            broadcastKeyHex = "cd".repeat(32),
            authorDid = "did:dht:z6MkAuthor2",
        )

        val args = stubBindings.lastEnableSiteProjectionArgs
        assertEquals(null, args?.indexPath)
        assertEquals(null, args?.maxAssetsPerDeploy)
        assertEquals(null, args?.maxDeploySizeBytes)
        assertEquals(null, args?.deployRetentionCount)
        assertEquals(null, args?.cspOverride)
    }

    // MARK: - commitDeploy

    @Test
    fun `commitDeploy returns asset count`() = runTest(testDispatcher) {
        val node = createNode()
        stubBindings.commitDeployResult = 42

        val count = node.commitDeploy("ctx-123", "deploy-abc")

        assertEquals(42, count)
    }

    @Test
    fun `commitDeploy propagates errors`() = runTest(testDispatcher) {
        val node = createNode()
        stubBindings.commitDeployError = BridgeException("not projected", "SCP-CTX-2080")

        assertFailsWith<BridgeException> {
            node.commitDeploy("ctx-bad", "deploy-xyz")
        }.also {
            assertTrue(it.message!!.contains("not projected"))
        }
    }

    // MARK: - rollbackDeploy

    @Test
    fun `rollbackDeploy delegates to bindings`() = runTest(testDispatcher) {
        val node = createNode()

        node.rollbackDeploy("ctx-123", "deploy-old")

        assertEquals("ctx-123", stubBindings.lastRollbackContextId)
        assertEquals("deploy-old", stubBindings.lastRollbackDeployId)
    }

    @Test
    fun `rollbackDeploy propagates errors`() = runTest(testDispatcher) {
        val node = createNode()
        stubBindings.rollbackDeployError = BridgeException("deploy not found", "SCP-CTX-2081")

        assertFailsWith<BridgeException> {
            node.rollbackDeploy("ctx-bad", "deploy-nope")
        }.also {
            assertTrue(it.message!!.contains("deploy not found"))
        }
    }

    // MARK: - disableSiteProjection

    @Test
    fun `disableSiteProjection delegates to bindings`() = runTest(testDispatcher) {
        val node = createNode()

        node.disableSiteProjection("ctx-123")

        assertEquals("ctx-123", stubBindings.lastDisableContextId)
    }

    @Test
    fun `disableSiteProjection is idempotent on unprojected context`() = runTest(testDispatcher) {
        val node = createNode()
        // Should not throw -- stub does nothing.
        node.disableSiteProjection("ctx-nonexistent")
    }

    // MARK: - Node properties

    @Test
    fun `node properties from handle JSON`() = runTest(testDispatcher) {
        val node = createNode()
        assertEquals("ws://127.0.0.1:9876/scp/v1", node.relayUrl)
        assertEquals(9876, node.relayPort)
        assertEquals("did:dht:z6MkTestNode", node.did)
    }

    // MARK: - Shutdown is one suspending path

    // Neither type implements AutoCloseable. A synchronous `close()` can only reach
    // `shutdown()` by blocking its calling thread on a coroutine whose dispatcher
    // neither type controls, which is what `.docs/lessons/kotlin/
    // oncleared-must-not-block-its-caller.md` forbids. Restoring `: AutoCloseable`
    // on either declaration fails this method.
    @Test
    fun `neither Relay nor Node implements AutoCloseable`() {
        assertFalse(
            AutoCloseable::class.java.isAssignableFrom(Relay::class.java),
            "Relay must expose suspend shutdown() alone, never a blocking close()",
        )
        assertFalse(
            AutoCloseable::class.java.isAssignableFrom(Node::class.java),
            "Node must expose suspend shutdown() alone, never a blocking close()",
        )
    }

    // Every declared method that stops a Node or a Relay suspends. A non-suspending
    // stop method under any name — `close`, `stop`, `dispose` — reintroduces that
    // same blocking shape, so this method matches on behaviour rather than on one
    // interface name. Kotlin compiles a suspend function to a JVM method taking a
    // trailing `kotlin.coroutines.Continuation`, which is how it is recognised here.
    @Test
    fun `every stop method on Relay and Node suspends`() {
        for (type in listOf(Relay::class.java, Node::class.java)) {
            val stopMethods = type.declaredMethods.filter { method ->
                method.name.substringBefore('$') in setOf("shutdown", "close", "stop", "dispose")
            }
            assertTrue(
                stopMethods.isNotEmpty(),
                "${type.simpleName} must declare a stop method",
            )
            for (method in stopMethods) {
                assertTrue(
                    method.parameterTypes.lastOrNull()?.name == "kotlin.coroutines.Continuation",
                    "${type.simpleName}.${method.name} must suspend, so a caller picks its dispatcher",
                )
            }
        }
    }

    @Test
    fun `shutdown marks a node shut down`() = runTest(testDispatcher) {
        val node = createNode()
        assertFalse(node.isShutdown)
        node.shutdown()
        assertTrue(node.isShutdown)
    }
}

// ---------------------------------------------------------------------------
// Stub ServerBindings
// ---------------------------------------------------------------------------

/**
 * Configurable stub for [ServerBindings] that records call arguments
 * and returns canned responses.
 */
internal class StubServerBindings : ServerBindings {

    // serve / httpUrl
    var serveResult: String = "127.0.0.1:8443"
    override fun nodeServe(handleJson: String, bindAddr: String?): String = serveResult

    var httpUrlResult: String? = null
    override fun nodeHttpUrl(handleJson: String): String? = httpUrlResult

    // Startup/shutdown
    var nodeStartInMemoryResult: String =
        """{"relayUrl":"ws://127.0.0.1:0/scp/v1","relayPort":0,"did":"did:dht:stub"}"""
    var relayStartInMemoryResult: String =
        """{"relayUrl":"ws://127.0.0.1:0/scp/v1","relayPort":0}"""

    override fun relayStartInMemory(): String = relayStartInMemoryResult
    override fun relayStartLocal(dataDir: String): String = relayStartInMemoryResult
    override fun nodeStartInMemory(identityDid: String?): String = nodeStartInMemoryResult
    override fun nodeStartLocal(
        dataDir: String,
        identityDid: String?,
        passphrase: String?,
    ): String = nodeStartInMemoryResult
    override fun relayShutdown(handleJson: String) { /* no-op stub */ }
    override fun nodeShutdown(handleJson: String) { /* no-op stub */ }

    // enableSiteProjection
    data class EnableArgs(
        val contextId: String,
        val broadcastKeyHex: String?,
        val authorDid: String?,
        val admission: String,
        val hostname: String,
        val indexPath: String?,
        val maxAssetsPerDeploy: Int?,
        val maxDeploySizeBytes: Long?,
        val deployRetentionCount: Int?,
        val cspOverride: String?,
    )

    var lastEnableSiteProjectionArgs: EnableArgs? = null

    @Suppress("LongParameterList")
    override fun nodeEnableSiteProjection(
        handleJson: String,
        contextId: String,
        admission: String,
        hostname: String,
        broadcastKeyHex: String?,
        authorDid: String?,
        indexPath: String?,
        maxAssetsPerDeploy: Int?,
        maxDeploySizeBytes: Long?,
        deployRetentionCount: Int?,
        cspOverride: String?,
    ) {
        lastEnableSiteProjectionArgs = EnableArgs(
            contextId, broadcastKeyHex, authorDid, admission, hostname,
            indexPath, maxAssetsPerDeploy, maxDeploySizeBytes, deployRetentionCount, cspOverride,
        )
    }

    // commitDeploy
    var commitDeployResult: Int = 0
    var commitDeployError: BridgeException? = null

    override fun nodeCommitDeploy(handleJson: String, contextId: String, deployId: String): Int {
        commitDeployError?.let { throw it }
        return commitDeployResult
    }

    // rollbackDeploy
    var lastRollbackContextId: String? = null
    var lastRollbackDeployId: String? = null
    var rollbackDeployError: BridgeException? = null

    override fun nodeRollbackDeploy(handleJson: String, contextId: String, deployId: String) {
        rollbackDeployError?.let { throw it }
        lastRollbackContextId = contextId
        lastRollbackDeployId = deployId
    }

    // disableSiteProjection
    var lastDisableContextId: String? = null

    override fun nodeDisableSiteProjection(handleJson: String, contextId: String) {
        lastDisableContextId = contextId
    }
}
